demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "cpu_monitor.h"
+#ifdef WIN32
+#include <tchar.h>
+
+namespace {
+std::size_t getNCores() {
+    SYSTEM_INFO sysinfo;
+    GetSystemInfo(&sysinfo);
+    return sysinfo.dwNumberOfProcessors;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::openQuery() {
+    PDH_STATUS status;
+
+    status = PdhOpenQuery(NULL, NULL, &query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhOpenQuery() failed");
+    }
+
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        TCHAR szCounterName[MAX_PATH];
+        _stprintf_s(szCounterName, sizeof(szCounterName), TEXT("\\Processor(%u)\\%% Processor Time"), i);
+
+        PDH_HCOUNTER counter;
+
+        status = PdhAddCounter(query, szCounterName, 0, &counter);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhAddCounter() failed");
+        }
+
+        status = PdhSetCounterScaleFactor(counter, -2); // scale counter to [0, 1]
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhSetCounterScaleFactor() failed");
+        }
+
+        coreTimeCounters.push_back(counter);
+    }
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status)
+    {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+}
+
+void CpuMonitor::closeQuery() {
+    PDH_STATUS status = PdhCloseQuery(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCloseQuery() failed");
```

Comment:
The only way we can get here is if `query` is an invalid handle, which shouldn't ever happen. So this should be a `logic_error`.

Reply:
I'm not sure this is the only case (how do you know?). May be something happens and fails internally. So should it be `system_error` too?

Reply:
> how do you know?

There's a list (of one entry) of possible errors in the [documentation](https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhclosequery).

> May be something happens and fails internally.

If something fails that shouldn't have, that _is_ a logic error. :-)

Reply:
OK

Reply:
updated